python/pyiceberg/expressions/base.py

```diff
@@ -672,3 +683,42 @@ def visit_unbound_predicate(self, predicate) -> BooleanExpression:
 
     def visit_bound_predicate(self, predicate) -> BooleanExpression:
         raise TypeError(f"Found already bound predicate: {predicate}")
+
+
+class BoundBooleanExpressionVisitor(BooleanExpressionVisitor[T], ABC):
+    @abstractmethod
+    def visit_in(self, ref: BoundReference[T], literals: set[Literal[Any]]) -> T:
+        """Visit the ref and literals from an In boolean expression
+        Args:
+            ref (BoundReference[T]): The ref used in the In boolean expression
+            literals (Tuple[Literal[T], ...]): The literals used in the In boolean expression
+        Raises:
+            NotImplementedError: If the concrete bound boolean expression visitor does not override this method
+        Returns:
+            R: The return type defined by the concrete bound boolean expression visitor
+        """
+
+    def visit_unbound_predicate(self, predicate: UnboundPredicate[T]):
+        """Visit an unbound predicate
+        Args:
+            predicate (UnboundPredicate[T]): An unbound predicate
+        Raises:
+            TypeError: This always raises since an unbound predicate is not expected in a bound boolean expression
+        """
+        raise TypeError(f"Not a bound predicate: {predicate}")
+
+    def visit_bound_predicate(self, predicate: BoundPredicate[T]) -> T:
```

Comment:
I think this should call another singledispatch method that accepts a bound predicate type and a bound expression visitor, rather than calling `isinstance` itself.